internal/tiltfile/live_update.go

```diff
@@ -0,0 +1,297 @@
+	return fmt.Sprintf("'%s'->'%s'", l.localPath, l.remotePath)
+}
+func (l liveUpdateSyncStep) Type() string { return "live_update_sync_step" }
+func (l liveUpdateSyncStep) Freeze()      {}
+func (l liveUpdateSyncStep) Truth() starlark.Bool {
+	return len(l.localPath) > 0 || len(l.remotePath) > 0
+}
+func (l liveUpdateSyncStep) Hash() (uint32, error) {
+	return starlark.Tuple{starlark.String(l.localPath), starlark.String(l.remotePath)}.Hash()
+}
+func (l liveUpdateSyncStep) liveUpdateStep()                      {}
+func (l liveUpdateSyncStep) declarationPosition() syntax.Position { return l.position }
+
+type liveUpdateRunStep struct {
+	command  string
+	triggers []string
+	position syntax.Position
+}
+
+var _ starlark.Value = liveUpdateRunStep{}
+var _ liveUpdateStep = liveUpdateRunStep{}
+
+func (l liveUpdateRunStep) String() string {
+	return fmt.Sprintf("command: %s ; triggers: %v", strconv.Quote(l.command), l.triggers)
+}
+func (l liveUpdateRunStep) Type() string { return "live_update_run_step" }
+func (l liveUpdateRunStep) Freeze()      {}
+func (l liveUpdateRunStep) Truth() starlark.Bool {
+	return len(l.command) > 0
+}
+func (l liveUpdateRunStep) Hash() (uint32, error) {
+	t := starlark.Tuple{starlark.String(l.command)}
+	for _, trigger := range l.triggers {
+		t = append(t, starlark.String(trigger))
+	}
+	return t.Hash()
+}
+func (l liveUpdateRunStep) declarationPosition() syntax.Position { return l.position }
+
+func (l liveUpdateRunStep) liveUpdateStep() {}
+
+type liveUpdateRestartContainerStep struct {
+	position syntax.Position
+}
+
+var _ starlark.Value = liveUpdateRestartContainerStep{}
+var _ liveUpdateStep = liveUpdateRestartContainerStep{}
+
+func (l liveUpdateRestartContainerStep) String() string                       { return "restart_container" }
+func (l liveUpdateRestartContainerStep) Type() string                         { return "live_update_restart_container_step" }
+func (l liveUpdateRestartContainerStep) Freeze()                              {}
+func (l liveUpdateRestartContainerStep) Truth() starlark.Bool                 { return true }
+func (l liveUpdateRestartContainerStep) Hash() (uint32, error)                { return 0, nil }
+func (l liveUpdateRestartContainerStep) declarationPosition() syntax.Position { return l.position }
+func (l liveUpdateRestartContainerStep) liveUpdateStep()                      {}
+
+func (s *tiltfileState) liveUpdateWorkDir(thread *starlark.Thread, fn *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var workDir string
+	if err := starlark.UnpackArgs(fn.Name(), args, kwargs, "dir", &workDir); err != nil {
+		return nil, err
+	}
+
+	if !filepath.IsAbs(workDir) {
+		return starlark.None, fmt.Errorf("work_dir path '%s' is not absolute", workDir)
+	}
+
+	ret := liveUpdateWorkDirStep{
+		value:    workDir,
+		position: thread.TopFrame().Position(),
+	}
+	s.unconsumedLiveUpdateSteps = append(s.unconsumedLiveUpdateSteps, ret)
```

Comment:
nice! ...is there some way we can have a common function that you use when implementing new Step types that does this automatically, so ppl writing this code don't forget to track the step here? (otherwise, maybe a big-ass note by where we define the types?)

Reply:
The only way I can think of to do this automatically is to wrap `s.builtin` to check the return value of every tiltfile function, and register any `liveUpdateStep`s it finds, but that feels gross. I've added a comment above `liveUpdateStep`.